util/fmtutil/format.go

```diff
@@ -0,0 +1,147 @@
+	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/prompb"
+)
+
+var MetricMetadataTypeValue = map[string]int32{
+	"UNKNOWN":        0,
+	"COUNTER":        1,
+	"GAUGE":          2,
+	"HISTOGRAM":      3,
+	"GAUGEHISTOGRAM": 4,
+	"SUMMARY":        5,
+	"INFO":           6,
+	"STATESET":       7,
+}
+
+// FormatMetrics convert metric family to a writerequest.
+func FormatMetrics(mf map[string]*dto.MetricFamily, extraLabels map[string]string) (*prompb.WriteRequest, error) {
+	wr := &prompb.WriteRequest{}
+
+	// build metric list
+	sortedMetricNames := make([]string, 0, len(mf))
+	for metric := range mf {
+		sortedMetricNames = append(sortedMetricNames, metric)
+	}
+	// sort metrics name in lexicographical order
+	sort.Strings(sortedMetricNames)
+
+	for _, metricName := range sortedMetricNames {
+		// Set metadata writerequest
+		mtype := MetricMetadataTypeValue[mf[metricName].Type.String()]
+		metadata := prompb.MetricMetadata{
+			MetricFamilyName: mf[metricName].GetName(),
+			Type:             prompb.MetricMetadata_MetricType(mtype),
+			Help:             mf[metricName].GetHelp(),
+		}
+		wr.Metadata = append(wr.Metadata, metadata)
+
+		for _, metric := range mf[metricName].Metric {
+			var timeserie prompb.TimeSeries
+
+			// build labels map
+			labels := make(map[string]string, len(metric.Label)+len(extraLabels))
+			labels[model.MetricNameLabel] = metricName
+
+			// add extra labels
+			for key, value := range extraLabels {
+				labels[key] = value
+			}
+
+			// add metric labels
+			for _, label := range metric.Label {
+				labelname := label.GetName()
+				if labelname == model.JobLabel {
+					labelname = fmt.Sprintf("%s%s", model.ExportedLabelPrefix, labelname)
+				}
+				labels[labelname] = label.GetValue()
+			}
+
+			// build labels name list
+			sortedLabelNames := make([]string, 0, len(labels))
+			for label := range labels {
+				sortedLabelNames = append(sortedLabelNames, label)
+			}
+			// sort labels name in lexicographical order
+			sort.Strings(sortedLabelNames)
+
+			for _, label := range sortedLabelNames {
+				timeserie.Labels = append(timeserie.Labels, prompb.Label{
+					Name:  label,
+					Value: labels[label],
+				})
+			}
+
+			timestamp := metric.GetTimestampMs()
+			if timestamp == 0 {
+				timestamp = time.Now().UnixNano() / int64(time.Millisecond)
+			}
+
+			timeserie.Samples = []prompb.Sample{
+				{
+					Timestamp: timestamp,
+					Value:     getMetricsValue(metric),
+				},
+			}
+
+			wr.Timeseries = append(wr.Timeseries, timeserie)
+		}
+	}
+	return wr, nil
+}
+
+// getMetricsValue return the value of a timeserie without the need to give value type
+func getMetricsValue(m *dto.Metric) float64 {
+	switch {
+	case m.Gauge != nil:
+		return m.GetGauge().GetValue()
+	case m.Counter != nil:
+		return m.GetCounter().GetValue()
+	case m.Untyped != nil:
+		return m.GetUntyped().GetValue()
+	default:
+		return 0.
```

Comment:
We should have more cases err and error in the default case

Reply:
I added summary and histogram support. From the [prometheus doc](https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format) only the following metric type are supported for text-based format: 
- Counter
- Gauge
- Histogram
- Summary
- Untyped

https://pkg.go.dev/github.com/prometheus/client_model/go#Metric

@roidelapluie I added summary and histogram support. From the [prometheus doc](https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format) only the following metric type are supported for text-based format: 
- Counter
- Gauge
- Histogram
- Summary
- Untyped

I also check the [io_prometheus_client library](https://pkg.go.dev/github.com/prometheus/client_model/go#Metric) and I didn't find the support for the following types:
- [GaugeHistogram](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#gaugehistogram-1)
- [Info](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#info-1)
- [StateSet](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#stateset-1)

Is there any reason why theses types are not (yet) supported by the prometheus_client library ?